File: windows/Scanner/RuleEvaluator.cs
using AIExposureScanner.Scanner.Models;
using AIExposureScanner.Scanner.Rules;

namespace AIExposureScanner.Scanner;

public sealed class RuleEvaluator
{
    public RuleEvaluator(IEnumerable<IRule>? rules = null)
    {
        Rules = (rules ?? DefaultRules).ToArray();
    }

    public IReadOnlyList<IRule> Rules { get; }

    public IReadOnlyList<Finding> Evaluate(ScanFacts facts) =>
        Rules
            .SelectMany(rule => rule.Evaluate(facts))
            .Distinct()
            .OrderBy(finding => finding.Severity.SortRank())
            .ThenBy(finding => finding.App, StringComparer.Ordinal)
            .ThenBy(finding => finding.ServerName ?? string.Empty, StringComparer.Ordinal)
            .ThenBy(finding => finding.ExtensionId ?? string.Empty, StringComparer.Ordinal)
            .ThenBy(finding => finding.RuleId, StringComparer.Ordinal)
            .ToArray();

    public static IReadOnlyList<IRule> DefaultRules { get; } =
    [
        new AESMCP001(),
        new AESMCP002(),
        new AESAUTH001(),
        new AESAUTH002(),
        new AESMCP003(),
        new AESMCP004(),
        new AESMCP005(),
        new AESEXT001(),
        new AESCFG001(),
        new AESMCP006(),
        new AESMCP007(),
        new AESAUTH003(),
        new AESCFG002(),
        new AESCFG003(),
        new AESCFG004()
    ];
}

File: macos/Scanner/Sources/Scanner/RuleEvaluator.swift
public struct RuleEvaluator: Sendable {
    public let rules: [any Rule]

    public init(rules: [any Rule] = RuleEvaluator.defaultRules) {
        self.rules = rules
    }

    public func evaluate(_ facts: ScanFacts) -> [Finding] {
        let findings = rules.flatMap { $0.evaluate(facts) }
        return Array(Set(findings)).sorted(by: FindingSort.areInIncreasingOrder)
    }

    public static let defaultRules: [any Rule] = [
        AESMCP001(),
        AESMCP002(),
        AESAUTH001(),
        AESAUTH002(),
        AESMCP003(),
        AESMCP004(),
        AESMCP005(),
        AESEXT001(),
        AESCFG001(),
        AESMCP006(),
        AESMCP007(),
        AESAUTH003(),
        AESCFG002(),
        AESCFG003(),
        AESCFG004(),
        AESMCP008(),
        AESMCP009(),
        AESMCP010(),
        AESMCP011(),
        AESMCP012(),
        AESMCP013(),
        AESAUTH004(),
        AESAUTH005(),
        AESAUTH006()
    ]
}

enum FindingSort {
    static func areInIncreasingOrder(_ lhs: Finding, _ rhs: Finding) -> Bool {
        let left = sortKey(lhs)
        let right = sortKey(rhs)
        return left < right
    }

    private static func sortKey(_ finding: Finding) -> String {
        [
            String(finding.severity.sortRank),
            finding.app,
            finding.serverName ?? "",
            finding.extensionId ?? "",
            finding.ruleId
        ].joined(separator: "|")
    }
}
